// @camunda8/sdk/dist/zeebe/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logging = exports.Types = exports.BpmnParser = exports.Duration = void 0;
var typed_duration_1 = require("typed-duration");
Object.defineProperty(exports, "Duration", { enumerable: true, get: function () { return typed_duration_1.Duration; } });
var BpmnParser_1 = require("./lib/BpmnParser");
Object.defineProperty(exports, "BpmnParser", { enumerable: true, get: function () { return BpmnParser_1.BpmnParser; } });
const SimpleLogger_1 = require("./lib/SimpleLogger");
const ZBJsonLogger_1 = require("./lib/ZBJsonLogger");
const ZBLogger_1 = require("./lib/ZBLogger");
const Interfaces = __importStar(require("./lib/interfaces-1.0"));
const GrpcInterfaces = __importStar(require("./lib/interfaces-grpc-1.0"));
const PublishedContract = __importStar(require("./lib/interfaces-published-contract"));
__exportStar(require("./zb/ZBWorker"), exports);
__exportStar(require("./zb/ZeebeGrpcClient"), exports);
__exportStar(require("./zb/ZeebeRESTClient"), exports);
/** @namespace */
exports.Types = {
    ...Interfaces,
    ...GrpcInterfaces,
    ...PublishedContract,
};
exports.Logging = {
    ZBJsonLogger: ZBJsonLogger_1.ZBJsonLogger,
    ZBLogger: ZBLogger_1.ZBLogger,
    ZBSimpleLogger: SimpleLogger_1.ZBSimpleLogger,
};
