// @camunda8/sdk/dist/c8/lib/C8Dto.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetVariableResponse = exports.SearchUsersResponse = exports.EvaluateDecisionResponse = exports.EvaluatedDecision = exports.EvaluatedInput = exports.MatchedRule = exports.EvaluatedOutput = exports.UploadDocumentsResponse = exports.UploadDocumentResponse = exports.PublishMessageResponse = exports.CorrelateMessageResponse = exports.BroadcastSignalResponse = exports.CreateProcessInstanceResponse = exports.DeployResourceResponse = exports.DeployResourceResponseDto = exports.FormDeployment = exports.DecisionRequirementsDeployment = exports.DecisionDeployment = exports.ProcessDeployment = exports.RestApiJob = void 0;
const lib_1 = require("../../lib");
class RestApiJob extends lib_1.LosslessDto {
}
exports.RestApiJob = RestApiJob;
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "jobKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "processInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "processDefinitionKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "elementInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "deadline", void 0);
class ProcessDeployment extends lib_1.LosslessDto {
}
exports.ProcessDeployment = ProcessDeployment;
__decorate([
    lib_1.Int64String
], ProcessDeployment.prototype, "processDefinitionKey", void 0);
class DecisionDeployment extends lib_1.LosslessDto {
}
exports.DecisionDeployment = DecisionDeployment;
__decorate([
    lib_1.Int64String
], DecisionDeployment.prototype, "decisionKey", void 0);
__decorate([
    lib_1.Int64String
], DecisionDeployment.prototype, "dmnDecisionRequirementsKey", void 0);
class DecisionRequirementsDeployment extends lib_1.LosslessDto {
}
exports.DecisionRequirementsDeployment = DecisionRequirementsDeployment;
__decorate([
    lib_1.Int64String
], DecisionRequirementsDeployment.prototype, "dmnDecisionRequirementsKey", void 0);
class FormDeployment {
}
exports.FormDeployment = FormDeployment;
__decorate([
    lib_1.Int64String
], FormDeployment.prototype, "formKey", void 0);
class DeployResourceResponseDto extends lib_1.LosslessDto {
}
exports.DeployResourceResponseDto = DeployResourceResponseDto;
__decorate([
    lib_1.Int64String
], DeployResourceResponseDto.prototype, "deploymentKey", void 0);
class DeployResourceResponse extends DeployResourceResponseDto {
}
exports.DeployResourceResponse = DeployResourceResponse;
class CreateProcessInstanceResponse {
}
exports.CreateProcessInstanceResponse = CreateProcessInstanceResponse;
__decorate([
    lib_1.Int64String
], CreateProcessInstanceResponse.prototype, "processDefinitionKey", void 0);
__decorate([
    lib_1.Int64String
], CreateProcessInstanceResponse.prototype, "processInstanceKey", void 0);
/** The signal was broadcast. */
class BroadcastSignalResponse extends lib_1.LosslessDto {
}
exports.BroadcastSignalResponse = BroadcastSignalResponse;
__decorate([
    lib_1.Int64String
    /** The unique ID of the signal that was broadcast. */
], BroadcastSignalResponse.prototype, "signalKey", void 0);
class CorrelateMessageResponse extends lib_1.LosslessDto {
}
exports.CorrelateMessageResponse = CorrelateMessageResponse;
__decorate([
    lib_1.Int64String
], CorrelateMessageResponse.prototype, "messageKey", void 0);
__decorate([
    lib_1.Int64String
], CorrelateMessageResponse.prototype, "processInstanceKey", void 0);
class PublishMessageResponse extends lib_1.LosslessDto {
}
exports.PublishMessageResponse = PublishMessageResponse;
__decorate([
    lib_1.Int64String
], PublishMessageResponse.prototype, "messageKey", void 0);
class UploadDocumentResponse extends lib_1.LosslessDto {
    constructor() {
        super(...arguments);
        /** Document discriminator. Always set to "camunda". */
        this['camunda.document.type'] = 'camunda';
    }
}
exports.UploadDocumentResponse = UploadDocumentResponse;
class UploadDocumentsResponse {
}
exports.UploadDocumentsResponse = UploadDocumentsResponse;
class EvaluatedOutput extends lib_1.LosslessDto {
}
exports.EvaluatedOutput = EvaluatedOutput;
class MatchedRule extends lib_1.LosslessDto {
}
exports.MatchedRule = MatchedRule;
__decorate([
    (0, lib_1.ChildDto)(EvaluatedOutput)
], MatchedRule.prototype, "evaluatedOutputs", void 0);
class EvaluatedInput extends lib_1.LosslessDto {
}
exports.EvaluatedInput = EvaluatedInput;
class EvaluatedDecision extends lib_1.LosslessDto {
}
exports.EvaluatedDecision = EvaluatedDecision;
__decorate([
    (0, lib_1.ChildDto)(MatchedRule)
], EvaluatedDecision.prototype, "matchedRules", void 0);
__decorate([
    (0, lib_1.ChildDto)(EvaluatedInput)
], EvaluatedDecision.prototype, "evaluatedInputs", void 0);
class EvaluateDecisionResponse extends lib_1.LosslessDto {
}
exports.EvaluateDecisionResponse = EvaluateDecisionResponse;
__decorate([
    (0, lib_1.ChildDto)(EvaluatedDecision)
], EvaluateDecisionResponse.prototype, "evaluatedDecisions", void 0);
class UserItem extends lib_1.LosslessDto {
}
__decorate([
    lib_1.Int64String
], UserItem.prototype, "id", void 0);
/** The user search result. */
class SearchUsersResponse extends lib_1.LosslessDto {
}
exports.SearchUsersResponse = SearchUsersResponse;
__decorate([
    (0, lib_1.ChildDto)(UserItem)
], SearchUsersResponse.prototype, "items", void 0);
class GetVariableResponse {
}
exports.GetVariableResponse = GetVariableResponse;
