// @camunda8/sdk/dist/operate/lib/OperateDto.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcessInstanceStatistics = exports.ChangeStatus = exports.Variable = exports.FlownodeInstance = exports.Incident = exports.ProcessInstance = exports.ProcessDefinition = exports.DecisionRequirements = exports.DecisionInstance = exports.DecisionInstanceOutput = exports.DecisionInstanceInput = exports.DecisionDefinition = void 0;
const lib_1 = require("../../lib");
class DecisionDefinition extends lib_1.LosslessDto {
}
exports.DecisionDefinition = DecisionDefinition;
__decorate([
    lib_1.Int64String
], DecisionDefinition.prototype, "key", void 0);
__decorate([
    lib_1.Int64String
], DecisionDefinition.prototype, "decisionRequirementsKey", void 0);
class DecisionInstanceInput extends lib_1.LosslessDto {
}
exports.DecisionInstanceInput = DecisionInstanceInput;
class DecisionInstanceOutput extends lib_1.LosslessDto {
}
exports.DecisionInstanceOutput = DecisionInstanceOutput;
class DecisionInstance extends lib_1.LosslessDto {
}
exports.DecisionInstance = DecisionInstance;
__decorate([
    lib_1.Int64String
], DecisionInstance.prototype, "key", void 0);
__decorate([
    lib_1.Int64String
], DecisionInstance.prototype, "processDefinitionKey", void 0);
__decorate([
    lib_1.Int64String
], DecisionInstance.prototype, "processInstanceKey", void 0);
__decorate([
    (0, lib_1.ChildDto)(DecisionInstanceOutput)
], DecisionInstance.prototype, "evaluatedOutputs", void 0);
class DecisionRequirements extends lib_1.LosslessDto {
}
exports.DecisionRequirements = DecisionRequirements;
__decorate([
    lib_1.Int64String
], DecisionRequirements.prototype, "key", void 0);
class ProcessDefinition extends lib_1.LosslessDto {
}
exports.ProcessDefinition = ProcessDefinition;
__decorate([
    lib_1.Int64String
], ProcessDefinition.prototype, "key", void 0);
class ProcessInstance extends lib_1.LosslessDto {
}
exports.ProcessInstance = ProcessInstance;
__decorate([
    lib_1.Int64String
], ProcessInstance.prototype, "key", void 0);
__decorate([
    lib_1.Int64String
], ProcessInstance.prototype, "parentKey", void 0);
__decorate([
    lib_1.Int64String
], ProcessInstance.prototype, "parentFlowNodeInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], ProcessInstance.prototype, "processDefinitionKey", void 0);
class Incident extends lib_1.LosslessDto {
}
exports.Incident = Incident;
__decorate([
    lib_1.Int64String
], Incident.prototype, "key", void 0);
__decorate([
    lib_1.Int64String
], Incident.prototype, "processDefinitionKey", void 0);
__decorate([
    lib_1.Int64String
], Incident.prototype, "processInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], Incident.prototype, "jobKey", void 0);
class FlownodeInstance extends lib_1.LosslessDto {
}
exports.FlownodeInstance = FlownodeInstance;
__decorate([
    lib_1.Int64String
], FlownodeInstance.prototype, "key", void 0);
__decorate([
    lib_1.Int64String
], FlownodeInstance.prototype, "processInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], FlownodeInstance.prototype, "incidentKey", void 0);
class Variable extends lib_1.LosslessDto {
}
exports.Variable = Variable;
__decorate([
    lib_1.Int64String
], Variable.prototype, "key", void 0);
__decorate([
    lib_1.Int64String
], Variable.prototype, "processInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], Variable.prototype, "scopeKey", void 0);
class ChangeStatus extends lib_1.LosslessDto {
}
exports.ChangeStatus = ChangeStatus;
__decorate([
    lib_1.Int64String
], ChangeStatus.prototype, "deleted", void 0);
class ProcessInstanceStatistics extends lib_1.LosslessDto {
}
exports.ProcessInstanceStatistics = ProcessInstanceStatistics;
__decorate([
    lib_1.Int64String
], ProcessInstanceStatistics.prototype, "active", void 0);
__decorate([
    lib_1.Int64String
], ProcessInstanceStatistics.prototype, "canceled", void 0);
__decorate([
    lib_1.Int64String
], ProcessInstanceStatistics.prototype, "incidents", void 0);
__decorate([
    lib_1.Int64String
], ProcessInstanceStatistics.prototype, "completed", void 0);
