// @camunda8/sdk/dist/admin/lib/AdminApiClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdminApiClient = void 0;
const debug_1 = __importDefault(require("debug"));
const got_1 = __importDefault(require("got"));
const lib_1 = require("../../lib");
const debug = (0, debug_1.default)('camunda:adminconsole');
/**
 * This class provides methods to interact with the Camunda Admin API.
 * @throws {RESTError} An error that may occur during API operations.
 */
class AdminApiClient {
    constructor(options) {
        const config = lib_1.CamundaEnvironmentConfigurator.mergeConfigWithEnvironment(options?.config ?? {});
        const prefixUrl = (0, lib_1.RequireConfiguration)(config.CAMUNDA_CONSOLE_BASE_URL, 'CAMUNDA_CONSOLE_BASE_URL');
        this.oAuthProvider =
            options?.oAuthProvider ?? (0, lib_1.constructOAuthProvider)(config);
        this.userAgentString = (0, lib_1.createUserAgentString)(config);
        this.rest = (0, lib_1.GetCustomCertificateBuffer)(config).then((certificateAuthority) => got_1.default.extend({
            prefixUrl,
            retry: lib_1.GotRetryConfig,
            https: {
                certificateAuthority,
            },
            handlers: [lib_1.beforeCallHook],
            hooks: {
                beforeRetry: [
                    (0, lib_1.makeBeforeRetryHandlerFor401TokenRetry)(this.getHeaders.bind(this)),
                ],
                beforeError: [(0, lib_1.gotBeforeErrorHook)(config)],
                beforeRequest: config.middleware ?? [],
            },
        }));
        debug('prefixUrl', `${prefixUrl}`);
    }
    async getHeaders() {
        const authorization = await this.oAuthProvider.getHeaders('CONSOLE');
        const headers = {
            'content-type': 'application/json',
            ...authorization,
            'user-agent': this.userAgentString,
            accept: '*/*',
        };
        return headers;
    }
    /**
     *
     * Get an array of the current API clients for this cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetClients) for more details.
     * @throws {RESTError}
     * @param clusterUuid - The cluster UUID
     *
     */
    async getClients(clusterUuid) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest(`clusters/${clusterUuid}/clients`, {
            headers,
        }).json();
    }
    /**
     * Create a new API client for a cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/CreateClient) for more details.
     * @throws {RESTError}
     */
    async createClient(req) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest
            .post(`clusters/${req.clusterUuid}/clients`, {
            body: JSON.stringify({
                clientName: req.clientName,
                permissions: req.permissions,
            }),
            headers,
        })
            .json();
    }
    /**
     * Get the details of an API client. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetClient) for more details.
     * @param clusterUuid
     * @param clientId
     * @throws {RESTError}
     * @returns
     */
    async getClient(clusterUuid, clientId) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest(`clusters/${clusterUuid}/clients/${clientId}`, {
            headers,
        }).json();
    }
    /**
     * See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/DeleteClient) for more details.
     * @param clusterUuid
     * @param clientId
     * @throws {RESTError}
     */
    async deleteClient(clusterUuid, clientId) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest
            .delete(`clusters/${clusterUuid}/clients/${clientId}`, {
            headers,
        })
            .json();
    }
    /**
     *
     * Return an array of clusters. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetClusters) for more details.
     * @throws {RESTError}
     */
    async getClusters() {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest('clusters', {
            headers,
        }).json();
    }
    /**
     *
     * Create a new cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/CreateCluster) for more details.
     * @throws {RESTError}
     */
    async createCluster(createClusterRequest) {
        const headers = await this.getHeaders();
        const req = {
            body: JSON.stringify(createClusterRequest),
            headers,
        };
        const rest = await this.rest;
        return rest.post('clusters', req).json();
    }
    /**
     *
     * Retrieve the metadata for a cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetCluster) for more details.
     * @throws {RESTError}
     *
     */
    async getCluster(clusterUuid) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest(`clusters/${clusterUuid}`, {
            headers,
        }).json();
    }
    /**
     *
     * Delete a cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/DeleteCluster) for more details.
     * @throws {RESTError}
     *
     */
    async deleteCluster(clusterUuid) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest
            .delete(`clusters/${clusterUuid}`, {
            headers,
        })
            .json();
    }
    /**
     *
     * Retrieve the available parameters for cluster creation. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetParameters) for more details.
     * @throws {RESTError}
     */
    async getParameters() {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest('clusters/parameters', {
            headers,
        }).json();
    }
    /**
     *
     * Retrieve the connector secrets. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/GetSecrets) for more details.
     * @throws {RESTError}
     */
    async getSecrets(clusterUuid) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest(`clusters/${clusterUuid}/secrets`, {
            headers,
        }).json();
    }
    /**
     *
     * Create a new connector secret. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/CreateSecret) for more details.
     * @throws {RESTError}
     */
    async createSecret({ clusterUuid, secretName, secretValue, }) {
        const headers = await this.getHeaders();
        const req = {
            body: JSON.stringify({ secretName, secretValue }),
            headers,
        };
        const rest = await this.rest;
        return rest.post(`clusters/${clusterUuid}/secrets`, req).json();
    }
    /**
     *
     * Delete a connector secret. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/DeleteSecret) for more details.
     * @throws {RESTError}
     */
    async deleteSecret(clusterUuid, secretName) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest
            .delete(`clusters/${clusterUuid}/secrets/${secretName}`, {
            headers,
        })
            .json();
    }
    /**
     *
     * Add one or more IPs to the whitelist for the cluster. See [the API Documentation](https://console.cloud.camunda.io/customer-api/openapi/docs/#/default/UpdateIpWhitelist) for more details.
     * @throws {RESTError}
     * @param ipwhitelist
     * @returns
     */
    async whitelistIPs(clusterUuid, ipwhitelist) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest
            .put(`clusters/${clusterUuid}/ipwhitelist`, {
            body: JSON.stringify({
                ipwhitelist,
            }),
            headers,
        })
            .json();
    }
    /**
     *
     * Retrieve a list of members and pending invites for your organisation. See the [API Documentation]() for more details.
     * @throws {RESTError}
     */
    async getUsers() {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest
            .get('members', {
            headers,
        })
            .json();
    }
    /**
     *
     * Add a member. See the [API Documentation]() for more details.
     * @throws {RESTError}
     *
     */
    async createMember(email, orgRoles) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest
            .post(`members/${email}`, {
            headers,
            body: JSON.stringify({ orgRoles }),
        })
            .json();
    }
    /**
     *
     * Delete a member from your organization. See the [API Documentation]() for more details.
     * @throws {RESTError}
     *
     */
    async deleteMember(email) {
        const headers = await this.getHeaders();
        const rest = await this.rest;
        return rest
            .delete(`members/${email}`, {
            headers,
        })
            .json();
    }
}
exports.AdminApiClient = AdminApiClient;
