// @camunda8/sdk/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Zeebe = exports.Tasklist = exports.Optimize = exports.Operate = exports.Modeler = exports.CamundaRestClient = exports.Camunda8 = exports.Auth = exports.Admin = exports.Dto = exports.HTTPError = exports.CamundaSDKConfiguration = exports.PollingOperation = void 0;
const Admin = __importStar(require("./admin"));
exports.Admin = Admin;
const c8_1 = require("./c8");
Object.defineProperty(exports, "Camunda8", { enumerable: true, get: function () { return c8_1.Camunda8; } });
const CamundaRestClient_1 = require("./c8/lib/CamundaRestClient");
Object.defineProperty(exports, "CamundaRestClient", { enumerable: true, get: function () { return CamundaRestClient_1.CamundaRestClient; } });
const lib_1 = require("./lib");
const CamundaSupportLogger_1 = require("./lib/CamundaSupportLogger");
const Modeler = __importStar(require("./modeler"));
exports.Modeler = Modeler;
const Auth = __importStar(require("./oauth"));
exports.Auth = Auth;
const Operate = __importStar(require("./operate"));
exports.Operate = Operate;
const Optimize = __importStar(require("./optimize"));
exports.Optimize = Optimize;
const Tasklist = __importStar(require("./tasklist"));
exports.Tasklist = Tasklist;
const Zeebe = __importStar(require("./zeebe"));
exports.Zeebe = Zeebe;
var PollingOperation_1 = require("./lib/PollingOperation");
Object.defineProperty(exports, "PollingOperation", { enumerable: true, get: function () { return PollingOperation_1.PollingOperation; } });
CamundaSupportLogger_1.CamundaSupportLogger.getInstance();
var lib_2 = require("./lib");
Object.defineProperty(exports, "CamundaSDKConfiguration", { enumerable: true, get: function () { return lib_2.CamundaSDKConfiguration; } });
var lib_3 = require("./lib");
Object.defineProperty(exports, "HTTPError", { enumerable: true, get: function () { return lib_3.HTTPError; } });
/**
 * These are decorators and infrastructure that are used to create Dtos for the Camunda 8 SDK.
 * A `LosslessDto` is used to represent data that is sent and received by the SDK - including job variables and headers.
 * They enable the lossless transformation of `int64` number values between the SDK and the Camunda 8 API —
 * representing the `int64` numbers as either `string` or `bigint` type in the SDK.
 *
 * See the {@link LosslessDto} for more information.
 */
exports.Dto = {
    ChildDto: lib_1.ChildDto,
    BigIntValue: lib_1.BigIntValue,
    BigIntValueArray: lib_1.BigIntValueArray,
    Int64String: lib_1.Int64String,
    Int64StringArray: lib_1.Int64StringArray,
    LosslessDto: lib_1.LosslessDto,
    createDtoInstance: lib_1.createDtoInstance,
};
